tsl/src/continuous_aggs/refresh.c

```diff
@@ -342,7 +357,37 @@ continuous_agg_refresh_internal(const ContinuousAgg *cagg,
 	 */
 	LockRelationOid(catalog_get_table_id(catalog, CONTINUOUS_AGGS_INVALIDATION_THRESHOLD),
 					AccessExclusiveLock);
-	continuous_agg_invalidation_threshold_set(cagg->data.raw_hypertable_id, refresh_window.end);
+
+	/* Compute new invalidation threshold. Note that this computation caps the
+	 * threshold at the end of the last bucket that holds data in the
+	 * underlying hypertable. */
+	computed_invalidation_threshold = invalidation_threshold_compute(cagg, &refresh_window);
+
+	/* Set the new invalidation threshold. Note that this only updates the
+	 * threshold if the new value is greater than the old one. Otherwise, the
+	 * existing threshold is returned. */
+	invalidation_threshold = invalidation_threshold_set_or_get(cagg->data.raw_hypertable_id,
+															   computed_invalidation_threshold);
+
+	/* We must also cap the refresh window at the invalidation threshold. If
+	 * we process invalidations after the threshold, the continuous aggregates
+	 * won't be refreshed when the threshold is moved forward in the
+	 * future. The invalidation threshold should already be aligned on bucket
+	 * boundary. */
+	if (invalidation_threshold <= refresh_window.end)
+	{
+		refresh_window.end = invalidation_threshold;
+
+		/* Capping the end might have made the window 0, or negative, so
+		 * nothing to refresh in that case */
+		if (refresh_window.start >= refresh_window.end)
+		{
+			emit_up_to_date_notice(cagg);
+			return;
+		}
```

Comment:
Will it be useful to check it anyway, i.e., outside the `if` statement?

Reply:
Sorry, what is it that you would like to check outside the if statement?

Reply:
Can `refresh_window.start >= refresh_window.end` happen if the window is not capped? In particular `refresh_window.start == refresh_window.end`

Reply:
Ohh, I think you want to check that start is not after end outside the larger if clause, or?

This is only an issue after capping since there is already a check above for the window given as arg to the function.

Reply:
I see. Since I suggested to use strict comparison between the window end and capped invalidation threshold, I wasn't sure if this nested if needs to be moved out or not.